libraries/libfc/src/crypto/ethereum/ethereum_utils.cpp

```diff
@@ -0,0 +1,208 @@
+#include <fc/crypto/ethereum/ethereum_utils.hpp>
+
+#include <ethash/keccak.hpp>
+
+namespace fc::crypto::ethereum {
+
+
+/**
+ * Removes "0x" prefix if present and converts hex string to lowercase
+ * @param hex The hex string to trim
+ * @return The trimmed and lowercase hex string
+ */
+std::string trim(const std::string& hex) {
+
+   std::string result = hex.starts_with("0x") ? hex.substr(2) : hex;
+   std::transform(result.begin(), result.end(), result.begin(), ::tolower);
+   return result;
+}
+
+std::string trim_public_key(const std::string& hex) {
+   auto clean_hex = trim(hex);
+   auto len       = clean_hex.length();
+
+   FC_ASSERT(std::ranges::contains(public_key_string_lengths, len), "Invalid public key length(${len}): ${hex}",
+             ("len", len)("hex", hex));
+
+   // If len mod 64 remainder is > 0, that means that the
+   // first two bytes denote the type and will be truncated
+   if (len % 64 > 0)
+      clean_hex = clean_hex.substr(2);
+
+   return clean_hex;
+}
+
+/**
+ * Converts a byte array to hexadecimal string
+ * @param bytes The byte array to convert
+ * @param add_prefix If true, adds "0x" prefix to the result
+ * @return The hexadecimal string representation
+ */
+std::string bytes_to_hex(const std::vector<uint8_t>& bytes, bool add_prefix) {
+   return (add_prefix ? "0x" : "") + fc::to_hex(bytes);
+}
+
+/**
+ * Converts a hexadecimal string to byte array
+ * @param hex The hexadecimal string to convert
+ * @return The resulting byte array
+ */
+std::vector<uint8_t> hex_to_bytes(const std::string& hex) {
+   auto clean_hex = trim(hex);
+   if (clean_hex.length() == 130) {
+      clean_hex = clean_hex.substr(2);
+   }
+
+   return fc::from_hex(clean_hex);
+}
+
+/**
+ * Creates an Ethereum-styled message hash from a payload string
+ * Uses the format: "\x19Ethereum Signed Message:\n" + length + message
+ * @param payload The message to hash
+ * @return The keccak256 hash of the formatted message
+ */
+em::message_hash_type hash_message(const em::message_body_type& payload) {
+
+   em::message_hash_type     eth_message_digest{};
+   std::vector<std::uint8_t> eth_message;
+   if (std::holds_alternative<std::vector<std::uint8_t>>(payload)) {
+      eth_message = std::get<std::vector<std::uint8_t>>(payload);
+   } else if (std::holds_alternative<std::string>(payload)) {
+      auto payload_str = std::get<std::string>(payload);
+
+      eth_message.resize(payload_str.size());
+      std::copy(payload_str.begin(), payload_str.end(), eth_message.begin());
+   } else {
+      auto payload_hash     = std::get<fc::sha256>(payload);
+      auto eth_message_size = payload_hash.data_size();
+      eth_message.resize(eth_message_size);
+
+      std::copy_n(payload_hash.data(),
+                  payload_hash.data_size(),
+                  eth_message.begin());
+
+   }
+
+   auto h256 = ethash::keccak256(eth_message.data(), eth_message.size());
+   std::copy_n(h256.bytes, sizeof(h256.bytes), eth_message_digest.data());
```

Comment:
It is not immediately clear to me that `eth_message_digest.data()` size will be `>= sizeof(h256.bytes)`.